rust/feature-flags/src/flag_matching.rs

```diff
@@ -805,6 +824,85 @@ impl FeatureFlagMatcher {
+        target_properties: &HashMap<String, Value>,
+    ) -> Result<bool, FlagError> {
+        Box::pin(self.evaluate_potentially_nested_cohort_filters(filters, target_properties)).await
+    }
+
+    async fn evaluate_potentially_nested_cohort_filters(
+        &self,
+        filters: &[PropertyFilter],
+        target_properties: &HashMap<String, Value>,
+    ) -> Result<bool, FlagError> {
+        let mut cohort_filters = Vec::new();
+        let mut non_cohort_filters = Vec::new();
+
+        // Separate cohort filters from non-cohort filters
+        for filter in filters {
+            if filter.prop_type == "cohort" {
+                cohort_filters.push(filter);
+            } else {
+                non_cohort_filters.push(filter);
+            }
+        }
+
+        // Evaluate non-cohort filters
+        for filter in &non_cohort_filters {
+            if !match_property(filter, target_properties, false).unwrap_or(false) {
+                return Ok(false);
+            }
+        }
+
+        // Evaluate cohort filters
+        if !cohort_filters.is_empty() {
+            let cohort_ids: HashSet<CohortId> = cohort_filters
+                .iter()
+                .filter_map(|f| f.value.as_i64().map(|id| id as CohortId))
+                .collect();
+
+            let cohorts = self.fetch_cohorts(cohort_ids.clone()).await?;
+            let seen_cohorts_cache: HashMap<CohortId, CohortOrEmpty> = cohorts
+                .into_iter()
+                .map(|cohort| (cohort.id, CohortOrEmpty::Cohort(cohort)))
+                .collect();
+
+            let sorted_cohort_ids = sort_cohorts_topologically(cohort_ids, &seen_cohorts_cache);
+
+            for cohort_id in sorted_cohort_ids {
+                if let Some(CohortOrEmpty::Cohort(cohort)) = seen_cohorts_cache.get(&cohort_id) {
+                    let cohort_property_filters = cohort.parse_filters()?; // TODO error handle
+                    let cohort_match = self
+                        .evaluate_cohort_filters(&cohort_property_filters, target_properties)
+                        .await?;
+
+                    let filter = cohort_filters
+                        .iter()
+                        .find(|f| f.value.as_i64() == Some(cohort_id as i64))
+                        .unwrap();
+                    match filter.operator {
+                        Some(OperatorType::In) if !cohort_match => return Ok(false),
+                        Some(OperatorType::NotIn) if cohort_match => return Ok(false),
+                        _ => {}
+                    }
+                }
+            }
+        }
+
+        Ok(true)
+    }
+
+    async fn fetch_cohorts(&self, cohort_ids: HashSet<CohortId>) -> Result<Vec<Cohort>, FlagError> {
+        let mut cohorts = Vec::new();
+        for &id in &cohort_ids {
+            let cohort = Cohort::from_pg(self.postgres_reader.clone(), id, self.team_id).await?;
+            cohorts.push(cohort);
+        }
+        Ok(cohorts)
+    }
```

Comment:
the risk here is that I'm calling this method a lot for cohorts with lots of dependencies.  This feels like it can be made more efficient.

Reply:
You should:
- Have a function that can take a `&[CohortID]` (or just a `TeamId`) and return a `Vec<Cohort>`
- Extremely aggressively cache this, probably even straight-up prefetching all cohorts for a given team

The N of cohorts per team is small on average, and the total number of bytes per cohort is small (~1KB-10KB on average I'd guess), so round-tripping to PG and executing the query a bunch of times is a lot worse than fetching more data than you need. The team with the most cohorts only has 10,000, and it falls off very rapidly (just checked). If cohorts are ~1KB each, that's a 10MB max per-team - if you set aside a 2-4GB to cache cohorts (with a cache ttl of, say, 5 minutes), you'll save PG a lot here.

You probably want to be clever and do a `select count` in the caching layer to decide whether to prefetch for a given team_id or fetch on a per-request basis, too.